core/Test/Tasty/Core.hs

```diff
@@ -361,22 +377,31 @@ foldTestTree
   -> TestTree
      -- ^ the tree to fold
   -> b
-foldTestTree (TreeFold fTest fGroup fResource fAfter) opts0 tree0 =
-  go mempty opts0 tree0
+foldTestTree (TreeFold fTest fGroup fAfter) opts0 tree0 =
+  snd (go mempty opts0 mempty tree0)
   where
-    go :: (Seq.Seq TestName -> OptionSet -> TestTree -> b)
-    go path opts tree1 =
+    go :: Seq.Seq TestName -> OptionSet -> Matched -> TestTree -> (Matched, b)
```

Comment:
I'm struggling with this: who precisely has `Matched`? Probably a good place to attach a comment, explaining the algorightm.

Reply:
Will do, but perhaps this will clarify it in the mean time:

```haskell
type TestMatched = Any
type ForceMatch = Any

go :: Seq.Seq TestName -> OptionSet -> ForceFilter -> TestTree -> (TestMatched, b)
```

Basically, when test `A` depends on `B`, and `A` is matched with `-p A` - `B` should also run. The code therefore keeps track of any matched tests, and forces their dependencies to match too (even if the filter would otherwise throw them out).

Reply:
Done